### PR TITLE
Lnd: Listen for RPC calls on a bufconn

### PR DESCRIPTION
**Pull request by orbitalturtle**
This PR adds a couple of pieces that can help us run Terminal in stateless mode, described here: https://github.com/lightninglabs/lightning-terminal/issues/160#issuecomment-888689859. 

In short, the first commit adds a field to ListenerCfg for passing in a bufconn to LND, which can be used to listen in-memory RPC calls. It also allows us to pass the admin macaroon back to the program calling LND via a channel.

The second commit adds a GetAllPermissions function that can be used by Terminal to retrieve all the permissions required by LND to bake a "super macaroon" that can be used to interact with lnd as well as Terminal's sub daemons. 

(I will also submit a PR to Terminal soon which will show how this all ties together.)

### CODE REVIEW

File: lnd.go
Comment (Roasbeef) [inline lnd.go:196]: Why isn't the existing `RPCListener` enough here?

Comment (Roasbeef) [reply lnd.go:196]: Looks like this is actually just to pass the macaroon over? So the existing one can be used?

Comment (orbitalturtle) [reply lnd.go:196]: Ah yeah we could. My reasoning for adding it as a separate field was mostly for clarity since we'll need both the buf and regular rpc listener & makes it easy to differentiate between them. But I suppose though we could just turn RPCListener into a slice and add both the regular listener and buf listener, then loop over the listeners inside, as needed? Is that what you're saying?

Comment (guggero) [reply lnd.go:196]: Yeah, I think that's what would be best since we turn that into a slice later on anyway: https://github.com/lightningnetwork/lnd/blob/master/lnd.go#L327

Comment (orbitalturtle) [reply lnd.go]: Sounds good, made that change

File: lnd.go
Comment (guggero) [inline lnd.go:175]: We should probably add a sentence here mentioning that the channel should be buffered to avoid lnd being blocked on sending to the channel.

File: lnd.go
Comment (guggero) [inline lnd.go:622]: The `range` command is nil-safe, so we don't need the extra `if`. And I think with the check if the mac channel is nil, we can just loop over `grpcListeners` directly.

File: mobile/bindings.go
Comment (guggero) [inline mobile/bindings.go:97]: I think the type definition can be omitted here since it's already in the slice declaration.
Also this can be made a bit more compact by using the `{{` notation (though that is subject to personal taste):

```
		RPCListeners: []*lnd.ListenerWithSignal{{
			Listener: lightningLis,
			Ready:    rpcListening,
		}},
```

File: lnd.go
Comment (guggero) [inline lnd.go:183]: Should probably change this to `if empty` and use `if len(lisCfg.RPCListeners) > 0 {` below.

File: rpcserver.go
Comment (guggero) [inline rpcserver.go:262]: I know we've been adding code in Pool and other repos for exporting this as well. But instead of exporting the individual permission components that exist, couldn't we just iterate over the map returned by `MainRPCServerPermissions()` and see what permissions are _actually used_ by any of the RPC methods?

**Review by guggero — COMMENTED** [review]
Thanks for the changes, code looks almost ready now!
Just a few more nits and questions.

File: rpcserver.go
Comment (orbitalturtle) [reply rpcserver.go:262]: Sure that sounds good to me. Loop over them in the Terminal code rather than here in lnd?

Comment (orbitalturtle) [reply rpcserver.go]: Sounds good, took a stab at that

File: rpcserver.go
Comment (guggero) [inline rpcserver.go:266]: This could be simplified by a double map (`map[string]map[string]struct{}`) that just map `entity -> action`. That way you could just insert into the map and in the end loop over it again to get de-duplicated values.

File: lnd.go
Comment (guggero) [inline lnd.go:627]: I think we should also close the mac channels after the `if !cfg.NoMacaroons {` block is over (where we also do `close(walletInitParams.MacResponseChan)`) to signal any listener that nothing is coming over that channel anymore. Especially for the case where no macaroons are used.

**Review by guggero — COMMENTED** [review]
Turned out great :100: 

Just one more comment, then we're good to go IMO.

**Comment by orbitalturtle** [comment]
@guggero Terrific, made those changes

**Comment by Roasbeef** [comment]
Needs a rebase!

File: rpcserver.go
Comment (Roasbeef) [inline rpcserver.go:264]: Style nit: missing a new line above here

File: rpcserver.go
Comment (Roasbeef) [inline rpcserver.go:279]: Can use do `struct{}{}` here instead of making the intermediate variable, same bleow.

File: rpcserver.go
Comment (Roasbeef) [inline rpcserver.go:284]: Style nit: extra new line

**Review by Roasbeef — APPROVED** [review]
LGTM ✨

Just a few style nits.

**Review by guggero — APPROVED** [review]
LGTM pending @Roasbeef's nits.

**Review by guggero — APPROVED** [review]

**Pull request merged**
